examples/keyset.rs

```diff
@@ -0,0 +1,524 @@
+    exit(1);
+}
+
+fn do_init(filename: &str, args: &[String]) {
+    if args.len() != 1 {
+        eprintln!("Usage: ... init <domain>");
+        exit(1);
+    }
+    let domainname = &args[0];
+
+    let ks = KeySet::new(Name::from_str(domainname).unwrap());
+
+    save_keyset(filename, &ks);
+}
+
+fn do_addkey(filename: &str, args: &[String]) {
+    if args.len() < 2 || args.len() > 3 {
+        eprintln!("Usage: ... addkey <key-type> <pubref> [<privref>]");
+        exit(1);
+    }
+    let keytype = &args[0];
+    let pubref = args[1].clone();
+    let privref = if args.len() == 3 {
+        Some(args[2].clone())
+    } else {
+        None
+    };
+
+    let mut ks = load_keyset(filename);
+    if keytype == "ksk" {
+        ks.add_key_ksk(pubref, privref, UnixTime::now());
+    } else if keytype == "zsk" {
+        ks.add_key_zsk(pubref, privref, UnixTime::now());
+    } else if keytype == "csk" {
+        ks.add_key_csk(pubref, privref, UnixTime::now());
+    } else {
+        eprintln!("Unknown key type '{keytype}'");
+        exit(1);
+    }
+    save_keyset(filename, &ks);
+}
+
+fn do_deletekey(filename: &str, args: &[String]) {
+    if args.len() != 1 {
+        eprintln!("Usage: ... deletekey <pubref>");
+        exit(1);
+    }
+    let pubref = &args[0];
+
+    let mut ks = load_keyset(filename);
+    if let Err(e) = ks.delete_key(pubref) {
+        eprintln!("Unable to delete key {pubref}: {e}");
+        exit(1);
+    }
+    save_keyset(filename, &ks);
+}
+
+fn do_start(filename: &str, args: &[String]) {
+    if args.len() != 1 {
+        eprintln!("Usage: ... start <roll-type>");
+        exit(1);
+    }
+    let rolltype = &args[0];
+
+    let rolltype = str_to_rolltype(rolltype);
+
+    let mut ks = load_keyset(filename);
+
+    // Find old and new keys.
+    let keys = ks.keys().to_vec();
+    let (old, new): (Vec<_>, Vec<_>) = keys
+        .into_iter()
+        .filter_map(|k: Key| match rolltype {
+            RollType::KskRoll => {
+                if let KeyType::Ksk(keystate) = k.keytype() {
+                    Some((keystate.clone(), k.pubref().to_string()))
+                } else {
+                    None
+                }
+            }
+            RollType::ZskRoll => {
+                if let KeyType::Zsk(keystate) = k.keytype() {
+                    Some((keystate.clone(), k.pubref().to_string()))
+                } else {
+                    None
+                }
+            }
+            RollType::CskRoll => match k.keytype() {
+                KeyType::Ksk(keystate)
+                | KeyType::Zsk(keystate)
+                | KeyType::Csk(keystate, _) => {
+                    Some((keystate.clone(), k.pubref().to_string()))
+                }
+                KeyType::Include(_) => None,
+            },
+        })
+        .filter(|(keystate, _)| !keystate.old())
```

Comment:
This is a bit confusing.  After this, you're partitioning the keys into new and old lists.  But you just filtered out "old" keys, right?  Perhaps a comment is necessary here explaining the difference between the two definitions of "old".

Reply:
I changed the comment to:
// Find old and new keys. First find the keys that have the right type for
// a roll type, and map to a keystate and a pubref. Filter out the keys
// are already old, we don't take them into account for a new key roll.
// Then split into two group: complete new keys (which are not
// signer, present, or at_parent) and old keys that do have one of
// those functions.

As we write signers, maybe some of this code needs to go to the library.